File: impl/python/tests/test_policy.py
# Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
"""C5 effect vocabulary and authorization conformance for the Python SDK (design.md §6),
graded against the shared independent corpus vectors/effect/cases.json (NOT produced by this
code): the full granted x effect authorization matrix (R-6.3), the principal-source gate --
only a signature-derived identity is an authorization principal, never transport metadata, a
foreign header, or a client-supplied name (R-6.5) -- and the optional signed safety label's
extraction from an object's ext map, including its malformed-rejection cases (R-6.4).

Run:  python -m unittest -v tests.test_policy      (from impl/python/)
"""
import json
import os
import unittest

from naalp import cbor, policy
from naalp.cbor import U, T, M


def _vectors():
    d = os.path.dirname(os.path.abspath(__file__))
    for _ in range(6):
        p = os.path.join(d, "vectors", "effect", "cases.json")
        if os.path.isfile(p):
            with open(p, encoding="utf-8") as f:
                return json.load(f)
        d = os.path.dirname(d)
    raise FileNotFoundError("vectors/effect/cases.json not found")


SRC_MAP = {
    "signature": policy.SOURCE_SIGNATURE,
    "transport_metadata": policy.SOURCE_TRANSPORT_METADATA,
    "foreign_header": policy.SOURCE_FOREIGN_HEADER,
    "client_name": policy.SOURCE_CLIENT_NAME,
}


class PolicyAuthorizationConformance(unittest.TestCase):
    C = _vectors()

    # ---- R-6.3: the full granted x effect authorization matrix ---------------------------

    def test_authorization_matrix(self):
        """TestAuthorizationMatrix (Go parity): a policy that ignored the object effect
        (constant allow/deny) would fail here -- the matrix contains both allows and
        denies."""
        rows = self.C["authorization_matrix"]
        self.assertEqual(len(rows), 16, "want 16 matrix cells")
        allows = denies = 0
        for r in rows:
            g = policy.Grant(principal="pA", max_effect=r["granted"])
            if r["allow"]:
                allows += 1
                # a granted row must SUCCEED and return None (no side effect, no error).
                self.assertIsNone(
                    g.authorize_object(policy.SOURCE_SIGNATURE, "pA", r["effect"]),
                    "granted=%d effect=%d: want allow" % (r["granted"], r["effect"]),
                )
            else:
                denies += 1
                with self.assertRaises(policy.PolicyError) as cm:
                    g.authorize_object(policy.SOURCE_SIGNATURE, "pA", r["effect"])
                self.assertEqual(
                    cm.exception.kind, "EffectNotAuthorized",
                    "granted=%d effect=%d" % (r["granted"], r["effect"]),
                )
            # the raw lattice must agree with the matrix.
            self.assertEqual(
                policy.authorizes(r["granted"], policy.normalize_effect(r["effect"])),
                r["allow"],
            )
        self.assertTrue(allows > 0 and denies > 0, "matrix must contain both allows and denies")

    # ---- R-6.5: only a signature-derived identity is an authorization principal ----------

    def test_principal_source_gate(self):
        """TestPrincipalSourceGate (Go parity): a grant that would authorize the effect is
        still denied when the presenter's identity is transport metadata, a foreign header,
        or a client-supplied name -- even for a read_only object within the ceiling."""
        rows = self.C["principal_sources"]
        self.assertTrue(len(rows) > 0)
        g = policy.Grant(principal="pA", max_effect=policy.DESTRUCTIVE)  # maximally permissive
        for row in rows:
            src = SRC_MAP[row["source"]]
            accepted = row["accepted"]

            if accepted:
                self.assertEqual(policy.resolve_auth_principal(src, "pA"), "pA")
            else:
                with self.assertRaises(policy.PolicyError) as cm:
                    policy.resolve_auth_principal(src, "pA")
                self.assertEqual(cm.exception.kind, "UnauthenticatedPrincipal", row["source"])

            # even a read_only object (within the ceiling) is denied from a non-signature
            # source.
            if accepted:
                self.assertIsNone(g.authorize_object(src, "pA", policy.READ_ONLY))
            else:
                with self.assertRaises(policy.PolicyError) as cm:
                    g.authorize_object(src, "pA", policy.READ_ONLY)
                self.assertEqual(cm.exception.kind, "UnauthenticatedPrincipal", row["source"])

    # ---- R-6.4: the optional signed safety label's extraction from an ext map ------------

    def test_safety_label_from_ext_present(self):
        sl = self.C["safety_label"]
        ext = M([(U(sl["ext_key"]), M([(U(1), T(sl["risk"])), (U(2), T(sl["scope"]))]))])
        label, present = policy.safety_label_from_ext(ext)
        self.assertTrue(present)
        self.assertEqual(label, policy.SafetyLabel(sl["risk"], sl["scope"]))
        # the built ext's inner map matches the independently-hex-pinned oracle bytes.
        inner = M([(U(1), T(sl["risk"])), (U(2), T(sl["scope"]))])
        self.assertEqual(cbor.encode(inner).hex(), sl["cbor_hex"])
        self.assertEqual(policy.SafetyLabel(sl["risk"], sl["scope"]).encode().hex(), sl["cbor_hex"])

    def test_safety_label_from_ext_absent(self):
        # an empty ext map: no key 1 present -> absent, no error.
        label, present = policy.safety_label_from_ext(M([]))
        self.assertIsNone(label)
        self.assertFalse(present)
        # no ext at all (None).
        label, present = policy.safety_label_from_ext(None)
        self.assertIsNone(label)
        self.assertFalse(present)

    def test_safety_label_from_ext_malformed_non_map(self):
        # ext[1] present but not a map at all.
        bad = M([(U(policy.SAFETY_LABEL_EXT_KEY), U(9))])
        with self.assertRaises(policy.PolicyError) as cm:
            policy.safety_label_from_ext(bad)
        self.assertEqual(cm.exception.kind, "MalformedSafetyLabel")

    def test_safety_label_from_ext_incomplete(self):
        # ext[1] is a map missing the scope field (key 2).
        bad = M([(U(policy.SAFETY_LABEL_EXT_KEY), M([(U(1), T("x"))]))])
        with self.assertRaises(policy.PolicyError) as cm:
            policy.safety_label_from_ext(bad)
        self.assertEqual(cm.exception.kind, "MalformedSafetyLabel")


if __name__ == "__main__":
    unittest.main()

File: impl/ruby/test/test_policy.rb
# Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
#
# C5 effect vocabulary and authorization conformance for the Ruby SDK (design.md §6; R-6.1..6.5),
# graded against the shared independent corpus vectors/effect/cases.json (NOT produced by this
# code): the four effect names + Bridge SafetyLabel identity mapping, unknown-value fail-closed
# normalization (R-6.2), the full granted×effect authorization matrix via Grant#authorize_object
# (R-6.3), the principal-source gate -- only a signature-derived identity is ever an authorization
# principal (R-6.5) -- and the optional signed safety label's CBOR bytes plus its
# present/absent/malformed ext-map handling (R-6.4). AuthorizeObject unconditionally returning OK
# (bypassing both the principal check and the ceiling check) is the C5 mutation target: it flips
# test_authorization_matrix and test_principal_source_gate on a denied-row assertion.
#
# Run:  ruby -Ilib -Itest test/test_policy.rb      (from impl/ruby/)
require 'minitest/autorun'
require 'json'
require 'naalp'

def effect_vectors
  d = File.expand_path(File.dirname(__FILE__))
  6.times do
    p = File.join(d, "vectors", "effect", "cases.json")
    return JSON.parse(File.read(p, encoding: "utf-8")) if File.file?(p)
    d = File.dirname(d)
  end
  raise "vectors/effect/cases.json not found"
end

SRC_MAP = {
  "signature" => Naalp::Policy::SOURCE_SIGNATURE,
  "transport_metadata" => Naalp::Policy::SOURCE_TRANSPORT_METADATA,
  "foreign_header" => Naalp::Policy::SOURCE_FOREIGN_HEADER,
  "client_name" => Naalp::Policy::SOURCE_CLIENT_NAME,
}.freeze

class PolicyConformance < Minitest::Test
  C = effect_vectors

  # ---- effect names + Bridge SafetyLabel mapping (byte/name parity) ---------------------

  def test_effect_names_and_bridge_match_oracle
    effs = C["effects"]
    assert_equal 4, effs.length
    effs.each do |e|
      assert_equal e["safety_label"], Naalp::Policy.safety_label_name(e["value"]), "effect #{e['value']} name"
    end
    C["bridge_mapping"].each do |m|
      assert_equal m["effect"], Naalp::Policy.normalize_effect(m["npamp_u8"]), "bridge u8 #{m['npamp_u8']}"
    end
  end

  # ---- R-6.2: an unrecognized effect value normalizes fail-closed to destructive --------

  def test_normalize_unknown_is_destructive
    (0..3).each { |v| assert_equal v, Naalp::Policy.normalize_effect(v) }
    unk = C["unknown_normalization"]
    refute_empty unk
    unk.each do |u|
      got = Naalp::Policy.normalize_effect(u["input"])
      assert_equal u["effect"], got, "normalize(#{u['input']})"
      assert_equal Naalp::Policy::DESTRUCTIVE, got, "unknown #{u['input']} must be destructive"
    end
  end

  # ---- R-6.3: the full granted x effect authorization matrix, via Grant#authorize_object --

  def test_authorization_matrix
    rows = C["authorization_matrix"]
    assert_equal 16, rows.length
    allows = 0
    denies = 0
    rows.each do |r|
      g = Naalp::Policy::Grant.new("pA", r["granted"])
      if r["allow"]
        allows += 1
        g.authorize_object(Naalp::Policy::SOURCE_SIGNATURE, "pA", r["effect"])
      else
        denies += 1
        err = assert_raises(Naalp::Policy::PolicyError) do
          g.authorize_object(Naalp::Policy::SOURCE_SIGNATURE, "pA", r["effect"])
        end
        assert_equal "EffectNotAuthorized", err.kind, "granted=#{r['granted']} effect=#{r['effect']}"
      end
      # the raw lattice must agree with the matrix
      assert_equal r["allow"],
                   Naalp::Policy.authorizes(r["granted"], Naalp::Policy.normalize_effect(r["effect"])),
                   "Authorizes(#{r['granted']},#{r['effect']})"
    end
    assert allows > 0 && denies > 0, "matrix must contain both allows and denies (allows=#{allows} denies=#{denies})"
  end

  # ---- R-6.5: only a signature-derived identity is ever an authorization principal ------

  def test_principal_source_gate
    g = Naalp::Policy::Grant.new("pA", Naalp::Policy::DESTRUCTIVE) # maximally permissive ceiling
    C["principal_sources"].each do |ps|
      src = SRC_MAP.fetch(ps["source"]) { flunk "unknown principal source #{ps['source']} in corpus" }

      if ps["accepted"]
        assert_equal "pA", Naalp::Policy.resolve_auth_principal(src, "pA"), ps["source"]
      else
        err = assert_raises(Naalp::Policy::PolicyError) { Naalp::Policy.resolve_auth_principal(src, "pA") }
        assert_equal "UnauthenticatedPrincipal", err.kind, ps["source"]
      end

      # even a read_only object (within the ceiling) is denied from a non-signature source.
      if ps["accepted"]
        g.authorize_object(src, "pA", Naalp::Policy::READ_ONLY)
      else
        err = assert_raises(Naalp::Policy::PolicyError) { g.authorize_object(src, "pA", Naalp::Policy::READ_ONLY) }
        assert_equal "UnauthenticatedPrincipal", err.kind, ps["source"]
      end
    end
  end

  # ---- R-6.4: the safety-label CBOR equals the independently constructed oracle bytes ---

  def test_safety_label_bytes_match_oracle
    sl = C["safety_label"]
    got = Naalp::Policy::SafetyLabel.new(sl["risk"], sl["scope"]).encode
    assert_equal sl["cbor_hex"], got.unpack1("H*")
    # module-level convenience wrapper produces the same bytes
    assert_equal sl["cbor_hex"], Naalp::Policy.safety_label_bytes(sl["risk"], sl["scope"]).unpack1("H*")
  end

  # ---- SafetyLabelFromExt: present / absent / malformed, fail-closed --------------------

  def test_safety_label_from_ext
    sl = C["safety_label"]
    assert_equal sl["ext_key"], Naalp::Policy::SAFETY_LABEL_EXT_KEY

    label = Naalp::Policy::SafetyLabel.new(sl["risk"], sl["scope"])
    got, present = Naalp::Policy.safety_label_from_ext(label.ext)
    assert present
    assert_equal label, got

    # absent: an empty ext map
    got2, present2 = Naalp::Policy.safety_label_from_ext(Naalp::CBOR::M.new([]))
    assert_nil got2
    refute present2

    # absent: a nil ext
    got3, present3 = Naalp::Policy.safety_label_from_ext(nil)
    assert_nil got3
    refute present3

    # malformed: ext[1] present but not a map
    bad = Naalp::CBOR::M.new([[Naalp::CBOR::U.new(Naalp::Policy::SAFETY_LABEL_EXT_KEY), Naalp::CBOR::U.new(9)]])
    err = assert_raises(Naalp::Policy::PolicyError) { Naalp::Policy.safety_label_from_ext(bad) }
    assert_equal "MalformedSafetyLabel", err.kind

    # malformed: ext[1] a map missing the scope field
    bad2 = Naalp::CBOR::M.new([[Naalp::CBOR::U.new(Naalp::Policy::SAFETY_LABEL_EXT_KEY),
                                 Naalp::CBOR::M.new([[Naalp::CBOR::U.new(1), Naalp::CBOR::T.new("x")]])]])
    err2 = assert_raises(Naalp::Policy::PolicyError) { Naalp::Policy.safety_label_from_ext(bad2) }
    assert_equal "MalformedSafetyLabel", err2.kind
  end
end
